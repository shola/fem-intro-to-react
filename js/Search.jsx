import React from 'react';
import shows from '../public/data';

export default function Search() {
    return (
        <pre>
            <code>
                {JSON.stringify(shows, null, 4)}
            </code>
        </pre>
    );
}
