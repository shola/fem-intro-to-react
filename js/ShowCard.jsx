// @flow

import React from 'react';
import { ShowPropsType, ShowStateType } from './ShowTypes';

/**
 * We must add keys to components that will need to render in different places
 * in the DOM. React knows how to re-use keyed components, instead of 
 * creating new ones and wasting resources.
 */
const ShowCard = (props: ShowPropsType): React.Element<any> => {
    const {
        title,
        imdbID,
        poster,
        year,
        description
    } = props;
    return (
        <div className="show-card" key={imdbID}>
            <img className="show-card-img" src={`public/img/posters/${poster}`} />
            <div className="show-card-text">
                <h3 className="show-card-title">{title}</h3>
                <h4 className="show-card-year">({year})</h4>
                <p className="show-card-description">{description}</p>
            </div>
        </div>
    );
};

export default ShowCard;
