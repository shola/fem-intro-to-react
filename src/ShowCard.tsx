import * as React from 'react';

interface ShowPropsType {
    title: string,
    imdbID: string,
    poster: string,
    year: string,
    description: string,
    trailer: string
}

interface ShowStateType {
    searchTerm: string
}

const ShowCard = (props: ShowPropsType) => {
    const {
        title,
        imdbID,
        poster,
        year,
        description
    } = props;
    return (
        <div className="show-card" key={imdbID}>
            <img className="show-card-img" src={`public/img/posters/${poster}`} />
            <div className="show-card-text">
                <h3 className="show-card-title">{title}</h3>
                <h4 className="show-card-year">({year})</h4>
                <p className="show-card-description">{description}</p>
            </div>
        </div>
    );
};

export default ShowCard;
