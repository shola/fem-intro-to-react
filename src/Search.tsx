import * as React from 'react';
import data from '../public/data.js';
import ShowCard from './ShowCard';
import Header from './Header';

interface ShowPropsType {
    title: string,
    imdbID: string,
    poster: string,
    year: string,
    description: string,
    trailer: string,
    route: any,
    shows: any,
    handleSearchTerm?: any
}

interface ShowStateType {
    searchTerm: string
}

/**
 * Private helper that decides which shows should be displayed on the search page
 */
function _showFilterer(state: ShowStateType) {
    // return function _showCheck(show: ShowPropsType): boolean {
    return function _showCheck(show: ShowPropsType) {
        const searchSpace = `${show.title} ${show.description}`.toUpperCase();
        const searchFor = state.searchTerm.toUpperCase();

        return searchSpace.indexOf(searchFor) >= 0;
    };
}

/**
 * Render the search page, now with a little bit of state.
 */

class Search extends React.Component<ShowPropsType, ShowStateType> {
    constructor(props: ShowPropsType) {
        super(props);
        this.handleSearchTermEvent = this.handleSearchTermEvent.bind(this);
        this.state = {
            searchTerm: ''
        };
    }
    handleSearchTermEvent(e: any) {
        this.setState({ searchTerm: e.target.value });
    }
    render() {
        return (
            <div className="container">
                <Header
                    showSearch={true}
                    handleSearchTermEvent={this.handleSearchTermEvent}
                    searchTerm={this.state.searchTerm}
                />
                <div className="shows">
                    {data.shows.filter(_showFilterer(this.state)).map(ShowCard)}
                </div>
            </div>
        );
    }
}

export default Search;
