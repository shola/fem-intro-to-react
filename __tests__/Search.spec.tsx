import * as React from 'react';
import * as ReactDOM from 'react-dom';
import * as Renderer from 'react-test-renderer';
import * as enzyme from 'enzyme';
import Search from '../src/Search';

import ShowCard from '../src/ShowCard';
import data from '../public/data.js';

describe('Search Page', () => {
    it('should be the same as the last snapshot', () => {
        const component = Renderer.create(<Search />);
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('should render all show cards when the search term is empty', () => {
        // this test may not be needed since snapshots check the number of children...
        const component = Renderer.create(<Search />);
        const tree = component.toJSON();
        const header = tree.children[0];
        const shows = tree.children[1];

        expect(shows.children.length).toBe(data.shows.length);
    });

    it('should filter show cards based on search term (case insensitive)', () => {
        const component = Renderer.create(<Search />);
        let tree = component.toJSON();
        let shows = tree.children[1];
        const header = tree.children[0];
        const searchInput = header.children[1];

        // Manually trigger callback with new state
        searchInput.props.onChange({ target: { value: 'House of Cards' } });
        // Re-render
        tree = component.toJSON();
        shows = tree.children[1];
        expect(tree).toMatchSnapshot();
        expect(shows.children.length).toBe(1);

        // Manually trigger callback with new state
        searchInput.props.onChange({ target: { value: 'house OF caRDS' } });
        // Re-render
        tree = component.toJSON();
        shows = tree.children[1];
        expect(tree).toMatchSnapshot();
        expect(shows.children.length).toBe(1);

        // Manually trigger callback with new state
        searchInput.props.onChange({ target: { value: 'House' } });
        // Re-render
        tree = component.toJSON();
        shows = tree.children[1];
        expect(tree).toMatchSnapshot();
        expect(shows.children.length).toBe(2);
    });
});
