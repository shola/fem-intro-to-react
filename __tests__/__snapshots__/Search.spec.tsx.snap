// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Page should be the same as the last snapshot 1`] = `
<div
  className="container"
>
  <header
    className="header"
  >
    <h1
      className="brand"
    >
      svideo
    </h1>
    <input
      className="search-input"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value=""
    />
  </header>
  <div
    className="shows"
  >
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/hoc.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          House of Cards
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2013–
          )
        </h4>
        <p
          className="show-card-description"
        >
          A Congressman works with his equally conniving wife to exact revenge on the people who betrayed him.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/dd.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Daredevil
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2015–
          )
        </h4>
        <p
          className="show-card-description"
        >
          A blind lawyer with his other senses superhumanly enhanced fights crime as a costumed superhero.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/oitnb.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Orange Is the New Black
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2013–
          )
        </h4>
        <p
          className="show-card-description"
        >
          The story of Piper Chapman, a woman in her thirties who is sentenced to fifteen months in prison after being convicted of a decade-old crime of transporting money for her drug-dealing girlfriend.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/ad.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Arrested Development
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2003–
          )
        </h4>
        <p
          className="show-card-description"
        >
          Level-headed son Michael Bluth takes over family affairs after his father is imprisoned. But the rest of his spoiled, dysfunctional family are making his job unbearable.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/uks.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Unbreakable Kimmy Schmidt
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2015–
          )
        </h4>
        <p
          className="show-card-description"
        >
          A woman is rescued from a doomsday cult and starts life over again in New York City.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/mp.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Marco Polo
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2014–
          )
        </h4>
        <p
          className="show-card-description"
        >
          In a world replete with greed, betrayal, sexual intrigue and rivalry, "Marco Polo" is based on the famed explorer's adventures in Kublai Khan's court in 13th century China.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/bl.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Bloodline
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2015–
          )
        </h4>
        <p
          className="show-card-description"
        >
          A family of adult siblings find that their secrets and scars are revealed when their black sheep of a brother returns home.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/bh.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          BoJack Horseman
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2014–
          )
        </h4>
        <p
          className="show-card-description"
        >
          Meet the most beloved sitcom horse of the '90s - 20 years later. BoJack Horseman was the star of the hit TV show "Horsin' Around," but today he's washed up, living in Hollywood, complaining about everything, and wearing colorful sweaters.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/hg.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Hemlock Grove
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2013–
          )
        </h4>
        <p
          className="show-card-description"
        >
          A teenage girl is brutally murdered, sparking a hunt for her killer. But in a town where everyone hides a secret, will they find the monster among them?
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/tk.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          The Killing
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2011–
          )
        </h4>
        <p
          className="show-card-description"
        >
          A police investigation, the saga of a grieving family, and a Seattle mayoral campaign all interlock after the body of 17-year-old Rosie Larsen is found in the trunk of a submerged car.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/l.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Love
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2016–
          )
        </h4>
        <p
          className="show-card-description"
        >
          A program that follows a couple who must navigate the exhilarations and humiliations of intimacy, commitment and other things they were hoping to avoid.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/mon.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Master of None
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2015–
          )
        </h4>
        <p
          className="show-card-description"
        >
          The personal and professional life of Dev, a 30-year-old actor in New York.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/f.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Flaked
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2016–
          )
        </h4>
        <p
          className="show-card-description"
        >
          Will Arnett will star as a self-help guru who needs a decent amount of help himself.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/jj.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Jessica Jones
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2015–
          )
        </h4>
        <p
          className="show-card-description"
        >
          A former superhero decides to reboot her life by becoming a private investigator.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/n.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Narcos
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2015–
          )
        </h4>
        <p
          className="show-card-description"
        >
          A chronicled look at the criminal exploits of Colombian drug lord Pablo Escobar.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/whas.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Wet Hot American Summer: First Day of Camp
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2015–
          )
        </h4>
        <p
          className="show-card-description"
        >
          Follow the counselors and campers on their first day at Camp Firewood in the summer of 1981.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/cdc.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Club de Cuervos
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2015–
          )
        </h4>
        <p
          className="show-card-description"
        >
          When the patriarch of a prominent family dies, his heirs battle to determine who will gain control of his beloved soccer team: The Cuervos of Nuevo Toledo.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/mam.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Making a Murderer
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2015–
          )
        </h4>
        <p
          className="show-card-description"
        >
          Filmed over a 10-year period, Making a Murderer is an unprecedented real-life thriller about Steven Avery, a DNA exoneree who, while in the midst of exposing corruption in local law enforcement, finds himself the prime suspect in a grisly new crime.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/s8.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Sense8
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2015–
          )
        </h4>
        <p
          className="show-card-description"
        >
          A group of people around the world are suddenly linked mentally, and must find a way to survive being hunted by those who see them as a threat to the world's order.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/fh.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Fuller House
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2016–
          )
        </h4>
        <p
          className="show-card-description"
        >
          D.J. Tanner-Fuller is a widow and mother of three. Things become too much to handle, so she asks for help from her sister Stephanie and her best friend Kimmy.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/wbd.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          w/ Bob & David
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2016–
          )
        </h4>
        <p
          className="show-card-description"
        >
          Bob and David emerge from the time machine they entered sixteen years ago and are reunited with the cast of Mr. Show.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Search Page should filter show cards based on search term (case insensitive) 1`] = `
<div
  className="container"
>
  <header
    className="header"
  >
    <h1
      className="brand"
    >
      svideo
    </h1>
    <input
      className="search-input"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value="House of Cards"
    />
  </header>
  <div
    className="shows"
  >
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/hoc.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          House of Cards
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2013–
          )
        </h4>
        <p
          className="show-card-description"
        >
          A Congressman works with his equally conniving wife to exact revenge on the people who betrayed him.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Search Page should filter show cards based on search term (case insensitive) 2`] = `
<div
  className="container"
>
  <header
    className="header"
  >
    <h1
      className="brand"
    >
      svideo
    </h1>
    <input
      className="search-input"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value="house OF caRDS"
    />
  </header>
  <div
    className="shows"
  >
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/hoc.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          House of Cards
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2013–
          )
        </h4>
        <p
          className="show-card-description"
        >
          A Congressman works with his equally conniving wife to exact revenge on the people who betrayed him.
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Search Page should filter show cards based on search term (case insensitive) 3`] = `
<div
  className="container"
>
  <header
    className="header"
  >
    <h1
      className="brand"
    >
      svideo
    </h1>
    <input
      className="search-input"
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value="House"
    />
  </header>
  <div
    className="shows"
  >
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/hoc.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          House of Cards
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2013–
          )
        </h4>
        <p
          className="show-card-description"
        >
          A Congressman works with his equally conniving wife to exact revenge on the people who betrayed him.
        </p>
      </div>
    </div>
    <div
      className="show-card"
    >
      <img
        className="show-card-img"
        src="public/img/posters/fh.jpg"
      />
      <div
        className="show-card-text"
      >
        <h3
          className="show-card-title"
        >
          Fuller House
        </h3>
        <h4
          className="show-card-year"
        >
          (
          2016–
          )
        </h4>
        <p
          className="show-card-description"
        >
          D.J. Tanner-Fuller is a widow and mother of three. Things become too much to handle, so she asks for help from her sister Stephanie and her best friend Kimmy.
        </p>
      </div>
    </div>
  </div>
</div>
`;
